File: src/commands/loyalty.js
// commands/loyalty.js — Loyalty points engine
import { rpc } from "@/lib/apiClient";

export const getLoyaltySettings = (storeId) =>
  rpc("get_loyalty_settings", { store_id: storeId });

// UpdateLoyaltySettingsDto: { store_id, points_per_naira?, naira_per_point_redemption?,
//                             min_redemption_points?, expiry_days?, is_active? }
export const updateLoyaltySettings = (payload) =>
  rpc("update_loyalty_settings", payload);

// Returns: { customer_id, points, naira_value }
export const getLoyaltyBalance = (customerId, storeId) =>
  rpc("get_loyalty_balance", { customer_id: customerId, store_id: storeId });

export const getLoyaltyHistory = (customerId, limit = 50) =>
  rpc("get_loyalty_history", { customer_id: customerId, limit });

// EarnPointsDto: { customer_id, store_id, transaction_id?, sale_amount }
export const earnPoints = (payload) =>
  rpc("earn_points", payload);

// RedeemPointsDto: { customer_id, store_id, transaction_id?, points }
export const redeemPoints = (payload) =>
  rpc("redeem_points", payload);

// AdjustPointsDto: { customer_id, store_id, points, notes? }
export const adjustPoints = (payload) =>
  rpc("adjust_points", payload);

export const expireOldPoints = (storeId) =>
  rpc("expire_old_points", { store_id: storeId });

File: src-tauri/src/models/loyalty.rs
// ============================================================================
// LOYALTY MODELS
// ============================================================================

use serde::{Deserialize, Serialize};
use rust_decimal::Decimal;
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct LoyaltySettings {
    pub store_id:                   i32,
    pub points_per_naira:           Decimal,
    pub naira_per_point_redemption: Decimal,
    pub min_redemption_points:      i32,
    pub expiry_days:                i32,
    pub is_active:                  bool,
    pub created_at:                 DateTime<Utc>,
    pub updated_at:                 DateTime<Utc>,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct LoyaltyTransaction {
    pub id:             i32,
    pub customer_id:    i32,
    pub store_id:       i32,
    pub transaction_id: Option<i32>,
    pub r#type:         String,
    pub points:         i32,
    pub balance_after:  i32,
    pub notes:          Option<String>,
    pub created_by:     Option<i32>,
    pub created_at:     DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct LoyaltyBalance {
    pub customer_id: i32,
    pub points:      i32,
    pub naira_value: Decimal,
}

// ── DTOs ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct UpdateLoyaltySettingsDto {
    pub store_id:                   i32,
    pub points_per_naira:           Option<f64>,
    pub naira_per_point_redemption: Option<f64>,
    pub min_redemption_points:      Option<i32>,
    pub expiry_days:                Option<i32>,
    pub is_active:                  Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct EarnPointsDto {
    pub customer_id:    i32,
    pub store_id:       i32,
    pub transaction_id: Option<i32>,
    pub sale_amount:    f64,
}

#[derive(Debug, Deserialize)]
pub struct RedeemPointsDto {
    pub customer_id:    i32,
    pub store_id:       i32,
    pub transaction_id: Option<i32>,
    pub points:         i32,
}

#[derive(Debug, Deserialize)]
pub struct AdjustPointsDto {
    pub customer_id: i32,
    pub store_id:    i32,
    pub points:      i32,
    pub notes:       Option<String>,
}
